pde_control_gym/src/environments1d/brain_tumor_env.py

```diff
@@ -0,0 +1,491 @@
+import numpy as np
+import gymnasium as gym
+from gymnasium import spaces
+from typing import Callable, Optional
+
+from pde_control_gym.src.environments1d.base_env_1d import PDEEnv1D
+
+class BrainTumor1D(PDEEnv1D):
+    r"""
+    Brain Tumor 1D PDE
+
+    This class implements the 1D Brain Tumor DPR PDE and inherits from the class :class:`PDEEnv1D`. Thus, for a full list of of arguments, first see the class :class:`PDEEnv1D` in conjunction with the arguments presented here
+
+    :param D: Diffusion coefficient in units (mm^2/day)
+    :param rho: Proliferation rate in units (1/day)
+    :param alpha: Radio-sensitivity parameter in units (Gy^-1)
+    :param alpha_beta_ratio: Radio-biologic parameter measuring fractionation sensitivity in units (Gy)
+    :param k: Carrying capacity in units (cells/mm^3)
+    :param t1_detection_radius: Radius of tumor at detection in units (mm)
+    :param t1_death_radius: Radius of tumor at patient death (mm)
+    :param reset_init_condition_func: Function that resets initial PDE condition :math:`u(x, 0)`
+    :param total_dosage: Total radiation dosage in units (Gy)
+    :param verbose: Toggles print statements
+    """
+    def __init__(self,
+                 D: float = 0.2,
+                 rho: float = 0.03,
+                 alpha: float = 0.04,
+                 alpha_beta_ratio: int = 10,
+                 k: float = 1e5,
+                 t1_detection_radius: int = 15,
+                 t1_death_radius: int = 35,
+                 reset_init_condition_func = None,
+                 total_dosage = None,
+                 verbose = True,
+                 **kwargs     
+    ):
+        super().__init__(**kwargs)
+        self.verbose = verbose
+
+        # Override base env 1d state size
+        self.nt = int(round(self.T/self.dt)+1)
+        self.nx = int(round(self.X/self.dx)+1)
+        self.u = np.zeros((self.nt, self.nx))
+        self.dosage_vs_time = np.zeros(self.nt)
+        self.xScale = np.linspace(0, self.X, self.nx) # index -> space mapping
+        if self.verbose:
+            print(f"nx: {self.nx}, nt: {self.nt}")
+            print(f"u.shape: {self.u.shape}")
+
+        # Override base env action and space to [0, 1]
+        self.action_space = spaces.Box(
+            np.full(1, 0, dtype="float32"),
+            np.full(1, 1, dtype="float32")
+        )
+        # Override base env observation space
+        self.observation_space = spaces.Box(
+            np.full(self.nx, 0, dtype="float64"),
+            np.full(self.nx, k, dtype="float64"),
+            dtype=np.float64
+        )
+
+        # Constant simulation parameters
+        self.T1_DETECTION_THRESHOLD = 0.8
+        self.T2_DETECTION_THRESHOLD = 0.16
+        self.reset_init_condition_func = reset_init_condition_func
```

Comment:
HARD RULE. Never should we have explicitly set constants in our code. This should be config parameters at worst